[FLINK-37925][table] Define VARIANT cast rules so a cast never alters the stored value (#28758)
apache/flink: docs/content.zh/docs/sql/reference/data-types.md

Because JSON has no literal for float, date, timestamp, or binary values, `PARSE_JSON` never
produces `FLOAT`, `DATE`, `TIMESTAMP`, `TIMESTAMP_LTZ`, or `BYTES`. A `VARIANT` can still hold
those kinds when built by other producers, such as the programmatic `VariantBuilder` API or format
conversions like Avro.

The JSON specification has no `NaN` or infinity literals, so `PARSE_JSON('NaN')`,
`PARSE_JSON('Infinity')`, and `PARSE_JSON('-Infinity')` fail, and `TRY_PARSE_JSON` returns `NULL`.
A `VARIANT` has no dedicated kind for these values. To keep one, store it as a JSON string and cast
it back out, for example `CAST(CAST(PARSE_JSON('"Infinity"') AS STRING) AS FLOAT)`.

A primitive-valued `VARIANT` can be converted to a scalar type with `CAST` or `TRY_CAST`. Numeric
targets accept any numeric value, so `PARSE_JSON('42')` casts to `INT`, `BIGINT`, or `DOUBLE`. A
value outside an integer or `DECIMAL` target's range fails `CAST` and returns `NULL` for
`TRY_CAST`. Other targets require the stored value to be of the matching kind, and a mismatch is
handled the same way. Casting a `VARIANT` to `CHAR`/`VARCHAR` extracts the scalar value, so a

> stored string is returned unquoted (for example `foo`), while objects and arrays return their JSON

Reviewer: did you double check this behavior with other vendors? sounds incorrect to me.
```
while objects and arrays return their JSON
```

Reviewer: this is what JSON_STRING should be good for.

Author: For objects, arrays, bytes we rely on the `Variant#toJson()` implementation. All other types (i.e., String) we convert to literal String. This is because for a case like `CAST(PARSE_JSON('foo') AS STRING))` we want to return a string literal `foo` and not a JSON String `"foo"`.

representation. Use `JSON_STRING` for the JSON representation, where a string stays quoted (for

> example `"foo"`). Casting to `TIME` is not supported.

Reviewer: incomplete list anyway.
```suggestion
example `"foo"`).
```

Author: Removed

{{< hint info >}}

> Unlike a regular numeric cast, casting a numeric `VARIANT` never wraps on overflow. To narrow a 

Reviewer: ```suggestion
Unlike a regular numeric cast, casting a numeric `VARIANT` never overflows. To narrow a 
```

Author: addressed

> `VARIANT` with wrap-around semantics, first cast it to a type that fits the stored value, then apply a 

Reviewer: ```suggestion
`VARIANT`, first cast it to a type that fits the stored value, then apply a 
```

Author: Addressed

regular narrowing cast:

```sql

> CAST(CAST(PARSE_JSON('1000') AS INT) AS TINYINT) -- returns -24 (wrap-around)

Reviewer: ```suggestion
CAST(CAST(PARSE_JSON('1000') AS INT) AS TINYINT) -- returns -24 (after overflow)
```

Author: Changed

```

> {{< /hint >}}

Reviewer 2: what is difference here from non VARIANT case?

flexibility to support highly nested and evolving schema.

`VARIANT` allows for deeply nested data structures, such as arrays within arrays, maps within maps,
or combinations of both. This capability makes `VARIANT` ideal for scenarios where data complexity
and nesting are significant.

`VARIANT` allows schema evolution, enabling the storage of data with changing or unknown schemas
without requiring upfront schema definition. For example, if a new field is added to the data, it
can be directly incorporated into the `VARIANT` data without modifying the table schema. This is
particularly useful in dynamic environments where schemas may evolve over time.

A `VARIANT` stores a single value of one of the following kinds: `NULL`, `BOOLEAN`, `TINYINT`,
`SMALLINT`, `INT`, `BIGINT`, `FLOAT`, `DOUBLE`, `DECIMAL` (up to precision 38), `STRING`, `DATE`,
`TIMESTAMP`, `TIMESTAMP_LTZ`, `BYTES`, or a nested array or object. `TIMESTAMP` and `TIMESTAMP_LTZ`

> are stored with microsecond precision and `DATE` as a day count. There is no `TIME` kind.

Reviewer: ```suggestion
are stored with microsecond precision.
```

Author: addressed

| Integer beyond 64-bit, up to 38 significant digits  | `DECIMAL`                                       |
| Decimal in plain notation, up to precision/scale 38 | `DECIMAL`                                       |
| Number in scientific notation, e.g. `1.5e3`         | `DOUBLE`                                        |
| Number exceeding 38 digits of precision or scale    | `DOUBLE`                                        |
| String                                              | `STRING`                                        |
| Array                                               | array (elements encoded by the same rules)      |
| Object                                              | object (values encoded by the same rules)       |

Because JSON has no literal for float, date, timestamp, or binary values, `PARSE_JSON` never
produces `FLOAT`, `DATE`, `TIMESTAMP`, `TIMESTAMP_LTZ`, or `BYTES`. A `VARIANT` can still hold
those kinds when built by other producers, such as the programmatic `VariantBuilder` API or format
conversions like Avro.

The JSON specification has no `NaN` or infinity literals, so `PARSE_JSON('NaN')`,

> `PARSE_JSON('Infinity')`, and `PARSE_JSON('-Infinity')` fail, and `TRY_PARSE_JSON` returns `NULL`.

Reviewer: also include the `1e400` example here.

Author: Added

A `VARIANT` has no dedicated kind for these values. To keep one, store it as a JSON string and cast
it back out, for example `CAST(CAST(PARSE_JSON('"Infinity"') AS STRING) AS FLOAT)`.

A primitive-valued `VARIANT` can be converted to a scalar type with `CAST` or `TRY_CAST`. Numeric

> targets accept any numeric value, so `PARSE_JSON('42')` casts to `INT`, `BIGINT`, or `DOUBLE`. A

Reviewer: ```suggestion
targets accept any wider numeric value, so `PARSE_JSON('42')` casts to `INT`, `BIGINT`, or `DOUBLE`. A
```

Author: Addressed

value outside an integer or `DECIMAL` target's range fails `CAST` and returns `NULL` for
`TRY_CAST`. Other targets require the stored value to be of the matching kind, and a mismatch is

> handled the same way. Casting a `VARIANT` to `CHAR`/`VARCHAR` extracts the scalar value, so a

Reviewer: Let's not go into the details of CHAR/VARCHAR at this line
```suggestion
handled the same way. Casting a `VARIANT` to `STRING` extracts the scalar value, so a
```

Author: Addressed

stored string is returned unquoted (for example `foo`). A variant that holds an object, array, or
binary value is not castable to a string; use `JSON_STRING` for its JSON representation instead
(there a string stays quoted, for example `"foo"`). A bounded `CHAR(n)`/`VARCHAR(n)` target is length-checked strictly, with no
padding or truncation: `VARCHAR(n)` accepts up to `n` characters and `CHAR(n)` requires exactly `n`,

> otherwise `CAST` fails and `TRY_CAST` returns `NULL`.

Reviewer: people should know how they work already
```suggestion
padding or truncation.
```

Author: Addressed

{{< hint info >}}
Unlike a regular numeric cast, casting a numeric `VARIANT` never overflows. To narrow a 
`VARIANT`, first cast it to a type that fits the stored value, then apply a 
regular narrowing cast:

```sql
CAST(CAST(PARSE_JSON('1000') AS INT) AS TINYINT) -- returns -24 (after overflow)
```

> {{< /hint >}}

Reviewer: I would vote for dropping this paragraph entirely. Or make it less prominent. A `hint info` is definitely too much. A sentence like `Numeric overflows are prohibited.` should be enough after: `targets accept any wider numeric value, so `PARSE_JSON('42')` casts to `INT`, `BIGINT`, or `DOUBLE`. Numeric overflows are prohibited.`

Author: Dropped
